### PR TITLE
Patch BigQueryClient to retry on network error

### PR DESCRIPTION
**Pull request by jamesmcm**
## Description

This patch retries by re-creating the client object if it fails (up to a retry limit), so that the client will not be timed out when calling `complete()` on the BigQueryTarget after a long job.

## Motivation and Context

At the moment sometimes long jobs can cause the BigQueryClient to timeout since the connection is created at initialisation of the target.

## Have you tested this? If so, how?

Yes, passing this patched client worked for jobs that would have timeout issues previously.

### CODE REVIEW

File: luigi/contrib/bigquery.py [hunk 1]
```diff
@@ -147,6 +158,12 @@ def dataset_exists(self, dataset):
                         fetched_location if fetched_location is not None else 'unspecified',
                         dataset.location))
 
+        except (TimeoutError, BrokenPipeError, IOError) as bq_connection_error:
+            if self.retry_count > self.retry_limit:
+                raise Exception(f"Exceeded max retries for BigQueryClient connection error: {bq_connection_error}") from bq_connection_error
+            self.retry_count += 1
+            self.__initialise_client()
+            self.dataset_exists(dataset)
```
Comment (hirosassa) [inline luigi/contrib/bigquery.py:166]: How about using retry decorator to simplify the codes?
The implementation below will be useful for you.
https://github.com/spotify/luigi/blob/master/luigi/contrib/gcs.py#L72-L81

Comment (jamesmcm) [reply luigi/contrib/bigquery.py:166]: Thanks, I'll make that change.

Comment (hirosassa) [reply luigi/contrib/bigquery.py]: @jamesmcm Thank you! Looks good to me, but flake8 notifies some formatting problems.

Comment (jamesmcm) [reply luigi/contrib/bigquery.py]: Thanks, I fixed that too.

**Comment by hirosassa** [comment]
This PR looks great! 
@dlstadther @Tarrasch @dataex plz take a look.

**Review by dlstadther — APPROVED** [review]
lgtm

**Pull request merged**